Unit tests for corsika output module

VictorBarbosaMartins opened the pull request:
Following #454, this PR adds the unit tests for the new corsika output module and is still ongoing (draft)
Please note that the merge is requested from this branch to the `corsika_output_mini_master` (not the main `master`)

VictorBarbosaMartins commented:
Hi @orelgueta. I finished the first implementation of the unit tests. Could you please have a look?

- Some features changed in the module, since I found some mistakes and bugs while conducting the tests. Please let me know if you need more input or some of the plots that helped me develop the code, e.g. plotting the rotation of the coordinates.
- I think the tests are covering everything that I implemented in the previous PR, but I might have missed something.
- We can indeed improve the tests, for instance by mocking tests (https://stackoverflow.com/questions/2665812/what-is-mocking), but I would like to avoid complicating it for now and maybe improve the test code in a later opportunity.
- If you want to run yourself the unit tests you will have to install boost-histogram and pyeventio in your environment, since I have not yet updated the dependencies in the containers. 
- In my computer all unit tests are passing (I have not checked yet but I do not expect that the integration tests were affected). We will confirm once we try to merge to main.
- The next step would be the application but before that I would like to merge what we have so far to `main` such that we are able proceed with the renaming/reordering of stuff as quickly as possible.

Please let me know if you have any questions or if would like to discuss about the changes in a call/in person.

Thanks!

orelgueta commented on simtools/util/general.py:
```diff
@@ -801,43 +802,54 @@ def convert_2D_to_radial_distr(xaxis, yaxis, hist2d, bin_size=50, max_dist=1000)
+    bins_step = (
+        2 * max_dist / num_bins
+    )  # in the 2D array, the positive and negative direction count.
+    warn = False
```
Instead of this `warn` you can simply register the warning inside the for loop and break. You save this non-elegant variable, an `if` call and going through the second axis if the first one already is problematic.

orelgueta replied:
But you will still fill the histogram. The `break` will just stop the for loop of the warning, not stop running all together.

VictorBarbosaMartins replied:
I see your point now. indeed, it is more elegant, thanks.

orelgueta commented on simtools/util/general.py:
```diff
@@ -849,8 +861,13 @@ def save_dict_to_file(dictionary, file_name):
     """
-    with open(file_name, "wb") as f:
-        pickle.dump(dictionary, f)
+    logger = logging.getLogger(__name__)
```
Same as for visualize, worth checking if we can call this once at the top.

orelgueta replied:
Yeah, check what I committed and see if you like it.

orelgueta commented on simtools/util/general.py:
```diff
@@ -849,8 +861,13 @@ def save_dict_to_file(dictionary, file_name):
-    with open(file_name, "wb") as f:
-        pickle.dump(dictionary, f)
+    logger = logging.getLogger(__name__)
+    if isinstance(file_name, str) and file_name[-4:] != ".yml":
```
See my comment about this elsewhere.

orelgueta commented on tests/conftest.py:
```diff
@@ -246,3 +248,24 @@ def telescope_north_test_file():
+
+
+@pytest.fixture
+def corsika_output_file_name_string():
```
I don't think you mean `string` here right? Is it actually a prefix? 
Why isn't there a suffix to this file name? (The answer to that might be made clear below, so I hope I remember to remove this question if that's the case.)

VictorBarbosaMartins replied:
I dont understand the question. 'string' is a suffix here (it comes after the function name). and the suffix indicates that the file returned by this function is a string and not yet Path.

orelgueta replied:
If you added `string` just to note that it's not a `Path`, then I think it's not necessary and you can remove it. The point about the prefix is related to my question about the suffix. I.e., the file name doesn't end with `.txt` or `.dat` or whatever. I was wondering why and if it is intentional. Perhaps this name is actually a file "`stem`" (see the pathlib `stem` function) to which you add different suffixes in various tests. In that case I would replace `string` with `stem`.

VictorBarbosaMartins replied:
Actually that is intentional because the test file has no extension. The output is in binary. Do you suggest adding an extension to the file name?

I cannot simply remove "_string" because of the fixture just below it, which would have the same name. Both fixtures are used in different places and are necessary.

orelgueta commented on tests/conftest.py:
```diff
@@ -246,3 +248,24 @@ def telescope_north_test_file():
+
+
+@pytest.fixture
+def corsika_output_instance_set_histograms(db, io_handler, corsika_output_file_name):
```
Since you don't have another instance, maybe we don't need the `set_histograms` in the name?

VictorBarbosaMartins replied:
we do have another instance which is used only in `test_corsika_output` file. one instance is for applying the set_histograms while the other stays fixed. this can be confusing because the behaviour might change if we set the histograms in a different way. Therefore, I created one instance for the case in which I need to set the histograms (and then I always set them) and another one for the cases in which we do not need to set them

orelgueta replied:
OK this is because I didn't look at `test_corsika_output` yet. Let me keep this unresolved for now and look at it first then.

orelgueta commented on tests/unit_tests/utils/test_general.py:
```diff
@@ -206,3 +211,54 @@ def check_results(x_to_test, y_to_test, x_right, y_right, angle):
+    )
+    difference = radial_edges[:-1] - distance_to_center_1D
+
+    assert pytest.approx(difference[:-1], 1) == 0  # last value deviates
```
Why do you have `, 1` in the `approx`? That essentially means this assertion would always pass cause you are setting `rel` option of `approx` to `1` (i.e., you allow a 100% error). It should be something like `pytest.approx(difference[:-1], 1e-6)`

VictorBarbosaMartins replied:
ok, I thought it is was similar to np.around but I am correcting this now then

VictorBarbosaMartins replied:
I found that for this specific line I only need to include the argument abs=1 and that is ok. I am fine with the accuracy of 1 meter. for the other pytest.approx ( in the other files, I will correct now)

VictorBarbosaMartins replied:
The step of the array is 1 meter, I would not expect any estimate to get better than that. if you check the value of the difference it is 0.2 ~ 0.5

orelgueta replied:
OK!

orelgueta commented on tests/unit_tests/utils/test_general.py:
```diff
@@ -206,3 +211,54 @@ def check_results(x_to_test, y_to_test, x_right, y_right, angle):
+    gen.convert_2D_to_radial_distr(
+        xaxis, yaxis, distance_to_center_2D, num_bins=4 * num_bins, max_dist=max_dist
+    )
+    msg = (
```
It's really enough to check that just part of the message is printed (2-3 words) =)

orelgueta commented on tests/unit_tests/utils/test_general.py:
```diff
@@ -206,3 +211,54 @@ def check_results(x_to_test, y_to_test, x_right, y_right, angle):
+        with open("test_file.yml") as file:
+            new_example_dict = yaml.load(file)
+            assert new_example_dict == example_dict
+        Path("test_file.yml").unlink()
```
You should use `tmp_test_directory` instead of writing to the current directory and then you don't need to worry about deleting the file either.

VictorBarbosaMartins replied:
good idea, thanks

orelgueta replied:
Unresolving just to remind myself to check tomorrow if that's the implementation we used in the past (can't remember).

orelgueta replied:
Yes that's how we did it in the past as well. Resolving again.

orelgueta commented on tests/unit_tests/utils/test_general.py:
```diff
@@ -206,3 +211,54 @@ def check_results(x_to_test, y_to_test, x_right, y_right, angle):
+        Path("test_file.yml").unlink()
+
+    # Path
+    path = Path("test_file_2.yml")
```
Here as well write to `tmp_test_directory` (you didn't delete the file in this case).

orelgueta reviewed (commented):
Looks good so far, but did not yet look at `test_corsika_ouput.py`, will continue later.

VictorBarbosaMartins commented on simtools/visualization/visualize.py:
```diff
@@ -748,9 +748,10 @@ def _kernel_plot_2D_photons(corsika_output_instance, property_name, log_z=False)
         "time_altitude": "Altitude of emission (km)",
         "num_photons_per_telescope": "Telescope index",
     }
+    _logger = logging.getLogger(__name__)
```
Please note that this is how we are doing in other places as well (see image below). The line `logging.getLogger(__name__) ` appears only in 3 functions in this file.

<img width="954" alt="Screenshot 2023-06-20 at 15 06 00" src="https://github.com/gammasim/gammasim-tools/assets/50831535/9af5b92b-51bc-47e5-b5c1-47d6d3216bec">

I don't understand from which import you want to get the logger. The function here in visualize might be used by different objects, which could have different loggers. I am in favour of keeping the logger in the original function.

orelgueta replied:
I meant to move this call to get the logger outside of the functions so that we need to do it only once. I just tested it and it should work exactly the same as it is now. It has the advantage that we remove this duplicate lines. I committed this change to this branch, see if you agree with it and if the tests run well. If you do not agree just revert back.

VictorBarbosaMartins replied:
ok, I agree with that. I would have a hard time agreeing to import a logger. thanks for the commit!

VictorBarbosaMartins commented on simtools/corsika/corsika_output.py:
```diff
@@ -242,17 +253,17 @@ def telescope_indices(self, telescope_new_indices):
+                telescope_new_indices = np.array([telescope_new_indices])
             for i_telescope in telescope_new_indices:
-                if not isinstance(i_telescope, int):
+                if not isinstance(i_telescope, (int, np.int32, np.int64)):
```
ok, thank you.

VictorBarbosaMartins commented on simtools/corsika/corsika_output.py:
```diff
@@ -271,22 +282,44 @@ def hist_config(self):
+        input_dict, input_yaml = None, None
+        if isinstance(input_config, dict):
+            input_dict = input_config
+        elif isinstance(input_config, (type(Path("")), str)):
```
There is no treatment in `collect_data_from_yaml_or_dict ` for the case in which `in_yaml` is integer for example. We are at the moment relying on the error which would be raised by this `with open(in_yaml) as file:`. Do you want me to add this check to  `collect_data_from_yaml_or_dict `?

orelgueta replied:
No, it's perfectly fine to get the error from `with open(in_yaml)`. We don't need to add a check ourselves.

orelgueta replied:
But you still have this `elif` I suggest to change it to a simple `else`.

VictorBarbosaMartins replied:
sorry, somehow it didnt make it to the commit. I changed it now

VictorBarbosaMartins commented on simtools/corsika/corsika_output.py:
```diff
@@ -410,11 +442,17 @@ def _create_regular_axes(self, label):
+        Parameters
+        ----------
+        individual_telescopes: bool
+            if False, the histograms are supposed to be filled for all telescopes.
```
Good point. i corrected the explanation now

VictorBarbosaMartins commented on simtools/corsika/corsika_output.py:
```diff
@@ -447,6 +485,8 @@ def _create_histograms(self):
     def _fill_histograms(self, photons, azimuth_angle=None, zenith_angle=None):
         """Fill all the histograms created by self._create_histogram with the information of the
          photons on the ground.
+         If the azimuth and zenith angles are given the Cherenkov photon's coordinates are given in
```
That is correct. I changed the text now

VictorBarbosaMartins commented on simtools/corsika/corsika_output.py:
```diff
@@ -570,11 +612,48 @@ def set_histograms(self, telescope_indices=None, individual_telescopes=False):
+        return self._individual_telescopes
+
+    @individual_telescopes.setter
+    def individual_telescopes(self, new_individual_telescopes):
```
ok, I added it to the definition

VictorBarbosaMartins commented on simtools/corsika/corsika_output.py:
```diff
@@ -570,11 +612,48 @@ def set_histograms(self, telescope_indices=None, individual_telescopes=False):
+        if new_individual_telescopes is None:
+            self._individual_telescopes = False
+        else:
+            if isinstance(new_individual_telescopes, bool):
```
so you propose that the setter function becomes simply `self._individual_telescopes = new_individual_telescopes` ?

orelgueta replied:
No, the `if new_individual_telescopes is None:` should stay. I would just remove the `if isinstance(new_individual_telescopes, bool):` and raising the type error. So a simple `else` would suffice.

VictorBarbosaMartins replied:
Ok, it simplified a lot the function, thanks

VictorBarbosaMartins commented on simtools/corsika/corsika_output.py:
```diff
@@ -708,17 +787,19 @@ def get_2D_num_photons_distr(self):
         -------
         numpy.array
-            Number of photons per event per telescope.
+            Number of photons per event per telescope in self.telescope_indices.
```
I removed the parameters description, since they were reminiscent of previous implementations.

The function returns the distribution (histogram) of the number of photons in various events. It is important to give one an idea of how many events are necessary, for a specific CORSIKA configuration, to reach a normal distribution. We will in the future want to compare distributions for different inputs but we want to be sure that the number of events simulated is enough for the comparison.

`get_2D_num_photons_distr` is different than the other histogram functions. While the others will lead to a number of histograms, depending on the flags `individual_telescopes` and `telescope_indices`, `get_2D_num_photons_distr` always returns the same shape. The list added in `visualize` is a trick to allow to use the same for loop for all the distributions. in this sense, all distributions are lists.

VictorBarbosaMartins commented on simtools/corsika/corsika_output.py:
```diff
@@ -874,13 +967,13 @@ def num_photons_per_event(self):
     def get_num_photons_distr(self, bins=50, range=None, event_or_telescope="event"):
         """
-        Get the distribution of photons per event.
+        Get the distribution of photons per event or per telescope, depending on the string
```
I actually consider the numpy function confusing, since we always think in `X and Y` when plotting and not in `Y and X`. Since it is well documented here and the function does not have the same name as the numpy one, I would prefer to keep this order, unless you have a strong opinion

orelgueta replied:
I don't have a strong opinion but I am not sure why you mention `x` and `y` when in this case the point is about the histogram values and its edges (in which case the order between the two is fairly arbitrary).

VictorBarbosaMartins replied:
I now inverted the order in the files. indeed, it is rather arbitrary and not a convention. Note that `ax.pcolormesh` is written as ` ax.pcolormesh(x_edges[i_hist], y_edges[i_hist], hist_values[i_hist]` and `ax.bar` also `        ax.bar(edges[i_hist][:-1], hist_values[i_hist],`

orelgueta replied:
Right, when creating a figure it's reversed (which also makes sense), but when binning the data it's `hist, edges`. So now look how nice and consistent it looks (while earlier it was reversed):
```
if event_or_telescope == "event":
   hist, edges = np.histogram(self.num_photons_per_event, bins=bins, range=range)
return hist, edges
```

VictorBarbosaMartins commented on simtools/corsika/corsika_output.py:
```diff
@@ -926,10 +1019,9 @@ def num_photons_per_telescope(self):
-            self._num_photons_per_telescope = np.sum(
-                np.array(self.num_photons_per_event_per_telescope), axis=0
-            )
+        self._num_photons_per_telescope = np.sum(
```
because I want to allow the user to change the main flags `individual_telescopes` and `telescope_indices` and run `set_histograms` again. in that case, `_num_photons_per_telescope ` has to be recalculated to consider the possibly different telescopes given by `telescope_indices`

orelgueta replied:
OK!

VictorBarbosaMartins commented on simtools/corsika/corsika_output.py:
```diff
@@ -942,14 +1034,14 @@ def total_num_photons(self):
         """
-        if self._total_num_photons is None:
-            self._total_num_photons = np.sum(self.num_photons_per_event)
+        self._total_num_photons = np.sum(self.num_photons_per_event)
```
same as above. the number of photons can change if you change the telescopes indices and re run the set_histograms

orelgueta replied:
OK!

VictorBarbosaMartins commented on simtools/util/general.py:
```diff
@@ -742,12 +744,11 @@ def rotate(x, y, rotation_angle_phi, rotation_angle_theta=0 * u.rad):
-    y_trans = np.cos(rotation_angle_theta) * (
-        x * np.sin(rotation_angle_phi) + y * np.cos(rotation_angle_phi)
-    )
+    x_trans = x * np.cos(rotation_around_y_axis) * np.cos(rotation_around_z_axis) - y * np.cos(
```
right, I returned to the factorized version now

VictorBarbosaMartins commented:
Hi @orelgueta, thanks for the first round of comments! I implemented the ones that I immediately agreed with you. Some conversations are still open and need further input from you. unit tests are passing. I will deal with the pytest.approx in the other test files already

orelgueta commented on simtools/corsika/corsika_output.py:
```diff
@@ -515,7 +555,7 @@ def _fill_histograms(self, photons, azimuth_angle=None, zenith_angle=None):
                 hist_num += 1
 
-    def set_histograms(self, telescope_indices=None, individual_telescopes=False):
+    def set_histograms(self, telescope_indices=None, individual_telescopes=None, hist_config=None):
```
OK, that makes sense.

orelgueta commented on simtools/util/general.py:
```diff
@@ -742,12 +744,11 @@ def rotate(x, y, rotation_angle_phi, rotation_angle_theta=0 * u.rad):
+    x_trans = x * np.cos(rotation_around_y_axis) * np.cos(rotation_around_z_axis) - y * np.cos(
+        rotation_around_y_axis
+    ) * np.sin(rotation_around_z_axis)
+    y_trans = x * np.sin(rotation_around_z_axis) + y * np.cos(rotation_around_z_axis)
```
But now only the second rotation is applied only to `x`. Are you sure about it?
I really didn't check this because it is not really a standard rotation so unless we sit together it would take a bit of time for me to figure out exactly what you need. I just want to point that out so you can confirm. We can also sit together and look at it if you wish.

VictorBarbosaMartins replied:
yes, because the Y coordinates do not change in a rotation around the Y axis. To derive this result, I multiplied the matrix of rotation around Z and the matrix of rotation around Y:
![image](https://github.com/gammasim/gammasim-tools/assets/50831535/8a0042c4-5673-4b57-ab14-f85d3ea3da4d)

One may ask why Y axis and not X axis. To answer that one has to look to the coodinate system definition in CORSIKA:
<img width="665" alt="Screenshot 2023-06-21 at 09 14 38" src="https://github.com/gammasim/gammasim-tools/assets/50831535/0e261b0e-9f7c-4395-b0e9-213756858045">
The rotation of the plane XY in order to align it the incoming direction of the particle is around the Y axis.

orelgueta replied:
Yeah that makes sense. The confusing part is that we don't keep the z coordinate so it looks like we're "missing" an axis in the second rotation. It's obvious that at some point we will need a 3-dimensional rotation and then this function will become even more complicated... =)

orelgueta commented on tests/unit_tests/corsika/test_corsika_output.py:
```diff
@@ -1,29 +1,750 @@
+
+    with pytest.raises(ValueError):
+        corsika_output_instance_set_histograms._get_hist_1D_projection("label_not_valid")
+        assert (
```
Caplog?

orelgueta commented on tests/unit_tests/corsika/test_corsika_output.py:
```diff
@@ -1,29 +1,750 @@
+    )  # 2nd tel, 2nd event
+
+
+def test_get_hist_1D_projection(corsika_output_instance_set_histograms):
```
May I suggest to move this test above all the others which call `_get_hist_1D_projection`? It would make more sense I think.

orelgueta commented on tests/unit_tests/corsika/test_corsika_output.py:
```diff
@@ -1,29 +1,750 @@
+    hist, edges = corsika_output_instance_set_histograms.get_num_photons_distr(
+        bins=87, range=None, event_or_telescope="telescope"
+    )
+    print(hist)
```
This is probably something you forgot to delete.

orelgueta reviewed (commented):
Went over `test_corsika_output.py`. Other than the comments we discussed in person, wrote a few here. Remember to pull before you make changes.

VictorBarbosaMartins commented on tests/unit_tests/corsika/test_corsika_output.py:
```diff
@@ -1,29 +1,750 @@
+        },
+    ]
+
+    azimuth_angle = None
```
They are used in the `_fill_histograms`. They have to be None because the idea of this test is to test this part of the code, in which no rotations are conducted:
```
            if azimuth_angle is None or zenith_angle is None:
                photon_x, photon_y = photons_info["x"], photons_info["y"]
            else:
                photon_x, photon_y = rotate(
                    photons_info["x"],
                    photons_info["y"],
                    azimuth_angle,
                    zenith_angle,
                )
```

orelgueta replied:
Yes but you can simply call `_fill_historgrams` without passing these variables and they will be set to `None` anyway cause that's the default. If you want to be explicit you could call
```
corsika_output_instance_fill._fill_histograms(photons, None, None)
```
or if you want to be super explicit
```
corsika_output_instance_fill._fill_histograms(photons=photons, azimuth_angle=None, zenith_angle=None)
```

Either way you don't need to define these variables here.

VictorBarbosaMartins commented on tests/unit_tests/corsika/test_corsika_output.py:
```diff
@@ -1,29 +1,750 @@
+
+
+def test_get_hist_2D_projection(corsika_output_instance):
+    label = "hist_non_existent"
```
Ok, I moved it down. I am changing to caplog as well

VictorBarbosaMartins commented:
> Went over `test_corsika_output.py`. Other than the comments we discussed in person, wrote a few here. Remember to pull before you make changes.

Thanks for reviewing it. I think I addressed all your comments. Please let me know if something is missing.

orelgueta reviewed (approved):
All looks good, thanks for implementing the changes!
